File: 2024/Go/solutions/day3.go
package solutions

import (
	"fmt"
	"log"
	"os"
    "errors"
)

func parseDo(data []byte, length int,idxPtr *int) (bool,error){
    idx := *idxPtr
    defer func(){
            *idxPtr = idx
    }()
    do := "do()"
    dont := "don't()"
    if idx+6 <= length && string(data[idx:idx+7]) == dont{
        return false,nil
    }
    if idx+3 <= length && string(data[idx:idx+4]) == do {
        return true,nil
    }
    return false,errors.New("Didn't find do/don't")
}

func parseMult(data []byte,length int,idxPtr *int) (int){
    idx := *idxPtr
    defer func(){
            *idxPtr = idx
    }()
    for idx < length{
        if (idx + 3 >= length) || data[idx+1] != 'u' || data[idx+2] != 'l' || data[idx+3] != '('{
            idx += 1
            break
        }
        curIdx := idx + 4
        num1 := 0
        mult := 10
        commaFound := false
        digitFound := false
        for curIdx < length{
            if data[curIdx] == ','{
                commaFound = true
                curIdx += 1
                break
            }
            if data[curIdx] < '0' || data[curIdx] > '9'{
                break
            }
            digitFound = true
            num1 = (num1*mult) + int(data[curIdx] - '0')
            curIdx += 1
        }
        if !commaFound || !digitFound || num1 > 999{
            idx = curIdx
            break
        }
        num2 := 0
        bracketFound := false
        digitFound = false
        for curIdx < length{
            if data[curIdx] == ')'{
                bracketFound = true
                curIdx += 1
                break
            }
            if data[curIdx] < '0' || data[curIdx] > '9'{
                break
            }
            digitFound = true
            num2 = (num2*mult) + int(data[curIdx] - '0')
            curIdx += 1
        }
        if !bracketFound || !digitFound || num2 > 999{
            idx = curIdx
            break
        }
        idx = curIdx
        return num1*num2
    }
    return 0
}

func Day3(){
    data,err := os.ReadFile("../input.txt")
    if err != nil{
        log.Fatalf("Error reading data: %v\n",err)
    }
    part_1_answer,part_2_answer := 0,0
    length := len(data)
    idx := 0
    val := 0
    do := true
    for idx < length{
        if data[idx] != 'm' && data[idx] != 'd'{
            idx += 1
            continue
        }
        if data[idx] == 'm'{
            val  = parseMult(data,length,&idx)
            part_1_answer += val
            if do{
                part_2_answer += val
            }
        }else{
            doOrDont,err := parseDo(data,length,&idx) 
            idx += 1
            if err != nil{
                continue
            }
            do = doOrDont
        }
    }
    fmt.Printf("Part 1: %v,Part 2: %v",part_1_answer,part_2_answer)

}

File: 2024/Zig/src/day3.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

fn parseMult(data: []u8, length: usize, idxPtr: *u32) u64 {
    var idx: u32 = idxPtr.*;
    defer {
        idxPtr.* = idx;
    }
    while (idx < length) {
        if (idx + 3 >= length or data[idx + 1] != 'u' or data[idx + 2] != 'l' or data[idx + 3] != '(') {
            idx += 1;
            break;
        }
        var num1: u64 = 0;
        var commaFound: bool = false;
        var digitFound: bool = false;
        idx = idx + 4;
        while (idx < length) {
            if (data[idx] == ',') {
                commaFound = true;
                idx += 1;
                break;
            }
            if (data[idx] < '0' or data[idx] > '9') {
                return 0;
            }
            digitFound = true;
            num1 = (num1 * 10) + (data[idx] - '0');
            idx += 1;
        }
        if (!digitFound or !commaFound or num1 > 999) {
            return 0;
        }
        var num2: u64 = 0;
        var bracketFound: bool = false;
        digitFound = false;
        while (idx < length) {
            if (data[idx] == ')') {
                bracketFound = true;
                idx += 1;
                break;
            }
            if (data[idx] < '0' or data[idx] > '9') {
                return 0;
            }
            digitFound = true;
            num2 = (num2 * 10) + (data[idx] - '0');
            idx += 1;
        }
        if (!bracketFound or !digitFound or num2 > 999) {
            return 0;
        }
        return num1 * num2;
    }
    return 0;
}

fn parseDo(data: []u8, length: usize, idxPtr: *u32) ?bool {
    const idx: u32 = idxPtr.*;
    if (idx + 6 <= length and std.mem.eql(u8, data[idx .. idx + 7], "don't()")) {
        return false;
    }
    if (idx + 3 <= length and std.mem.eql(u8, data[idx .. idx + 4], "do()")) {
        return true;
    }
    return null;
}

pub fn day3_sol(allocator: Allocator) !void {
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();
    var part1: u64 = 0;
    var part2: u64 = 0;
    const data = try file.reader().readAllAlloc(allocator, std.math.maxInt(u32));
    defer allocator.free(data);
    const length: usize = data.len;
    var idx: u32 = 0;
    var value: u64 = 0;
    var do: bool = true;
    while (idx < length) {
        if (data[idx] != 'm' and data[idx] != 'd') {
            idx += 1;
            continue;
        }
        if (data[idx] == 'm') {
            value = parseMult(data, length, &idx);
            part1 += value;
            if (do) {
                part2 += value;
            }
        } else {
            const doOrDont: ?bool = parseDo(data, length, &idx);
            if (doOrDont) |val| {
                do = val;
            }
            idx += 1;
        }
    }
    std.debug.print("Part 1 : {d}, Part 2 : {d}\n", .{ part1, part2 });
}
